xbmc/pvr/windows/GUIWindowPVRBase.cpp

```diff
@@ -587,8 +587,12 @@ bool CGUIWindowPVRBase::EditTimer(CFileItem *item)
     return false;
   }
 
-  if (ShowTimerSettings(timer) && !timer->GetTimerType()->IsReadOnly())
-    return g_PVRTimers->UpdateTimer(timer);
+  // clone the timer.
+  const CPVRTimerInfoTagPtr newTimer(new CPVRTimerInfoTag);
+  newTimer->UpdateEntry(timer);
+
+  if (ShowTimerSettings(newTimer) && !timer->GetTimerType()->IsReadOnly())
+    return g_PVRTimers->UpdateTimer(newTimer);
```

Comment:
This is actually an ancient bug that pops up in this context and not really easy to explain. Took me a while to understand.

Creating timers works the way that a new timer tag not known to CPVRTimers singleton is created which will be send to the addon. Addon signals timer changes. Kodi calls GetTimers at the addon and syncs local timer list with new list obtained via GetTimers. During sync the new timer will be identified inserted at CPVRTimers singleton. https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/timers/PVRTimers.cpp#L186 All good.

Updating timers before this change works different. There the timer tag known to the CPVRTimers singleton will be updated directly and immediately, even before timer update has been confirmed by the backend. This is inconstent compared to new timer handling and leads to other problems.

Example: I create a manual timer, lets say with start/end 14:00/18:00  that spans 3 epg events on a certain channel. <code>CPVRTimers::UpdateEntries</code> inserts a corresponding timer tag at CPVRTimers singleton. This tag gets associated with the 3 abovementioned epg tags. PVR Guide window reflects this (3 timer icons). Now I change the end time for the timer from 18:00 to 16:00. let's assume this leads to the timer no longer covering the 3rd epg event. Kodi calls UpdateTimer at addon, then. Due to the fact that the origin timer tag is changed directly, <code>CPVRTimers::UpdateEntries</code> triggerd by UpdateTimer has no performant way to determine the 3rd epg tag with the timer that must be removed. Remember, the associated tag already has the new end time set. The information that the timer before was set to end at 18:00 is lost. With this change, this is no longer a problem.

Another problem with existing implementation is that an updated event might not get correctly identifies as updated by <code>CPVRTimers::UpdateEntries</code>, because the item in existing timer list is logically identically with the one delivered with new timer list.

Uff. HTH?